### PR TITLE
ROS 2 canopen manager and roboteq driver refactor

### PR DESCRIPTION
**Pull request by KmakD**
### Description

This waits for [DriverState update PR](https://github.com/husarion/panther_ros/pull/392) to be merged.
This PR mains to create Abstraction for Driver including its implementation for Roboteq. Also, it refactors canopen class

### Modifications

- renamed canopen_controller to canopen_manager
- removed drivers from canopen_manager
- created driver, motor_driver abstraction
- created robote1 driver and roboteq motor driver implementations
- updated tests for roboteq and canopen class

### CODE REVIEW

File: panther_hardware_interfaces/src/panther_system/motors_controller/roboteq_driver.cpp [hunk 1]
```diff
@@ -41,17 +41,10 @@ struct CANopenObject
 // parameter names changed, but ids stayed the same, it will be better to just use ids directly
```
Comment (pawelirh) [inline panther_hardware_interfaces/src/panther_system/motors_controller/roboteq_driver.cpp:36]: Comment to line 32-36

- How about moving the `CANOpenObject` to `canopen_manager`?
- `CANopenObject` -> `CANOpenObject`

File: panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+
+TEST_F(TestRoboteqDriver, BootErrorVendorId)
+{
+  roboteqs_mock_->GetDriver()->SetOnReadWait<std::uint32_t>(0x1018, 1, 100000);
+  ASSERT_FALSE(TestBoot());
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  roboteqs_mock_->GetDriver()->SetOnReadWait<std::uint32_t>(0x1018, 1, 0);
+  ASSERT_TRUE(TestBoot());
+}
```
Comment (pawelirh) [inline panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp]: Why do we test twice on these tests? I assume that the error scenario is the first assertion. If so, it seems to be enough for a test case.
Can we somehowe name the argumenets provided to the funtion, so we can have better understanding of provided values?

File: panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+
+  roboteqs_mock_->GetDriver()->SetCurrent(DriverChannel::CHANNEL1, motor_1_current);
+  roboteqs_mock_->GetDriver()->SetCurrent(DriverChannel::CHANNEL2, motor_2_current);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(50));
+
+  panther_hardware_interfaces::MotorDriverState fb_motor_1 =
+    roboteq_driver_->GetMotorDriver(kMotor1Name)->ReadMotorDriverState();
+  panther_hardware_interfaces::MotorDriverState fb_motor_2 =
+    roboteq_driver_->GetMotorDriver(kMotor2Name)->ReadMotorDriverState();
```
Comment (pawelirh) [inline panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp]: What `fb_` stands for?

File: panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+
+TEST_F(TestRoboteqDriver, ReadRoboteqMotorStatesTimestamps)
+{
+  BootRoboteqDriver();
+
+  panther_hardware_interfaces::MotorDriverState fb_motor_1 =
+    roboteq_driver_->GetMotorDriver(kMotor1Name)->ReadMotorDriverState();
+
+  // based on publishing frequency in the Roboteq mock (10)
+  std::this_thread::sleep_for(std::chrono::milliseconds(10));
```
Comment (pawelirh) [inline panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp]: Is value 10ms correct? Don't u aimed 100ms? (I refere to 10Hz from the comment)

File: panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+  roboteqs_mock_->GetDriver()->SetBatteryCurrent2(battery_current_2);
+
+  roboteqs_mock_->GetDriver()->SetDriverFaultFlag(DriverFaultFlags::OVERHEAT);
+  roboteqs_mock_->GetDriver()->SetDriverScriptFlag(DriverScriptFlags::ENCODER_DISCONNECTED);
+  roboteqs_mock_->GetDriver()->SetDriverRuntimeError(
+    DriverChannel::CHANNEL1, DriverRuntimeErrors::LOOP_ERROR);
+  roboteqs_mock_->GetDriver()->SetDriverRuntimeError(
+    DriverChannel::CHANNEL2, DriverRuntimeErrors::SAFETY_STOP_ACTIVE);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(50));
```
Comment (pawelirh) [inline panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp:231]: What is the execution time of roboteq unit tests? I understand that these sleeps are needed to operate with roboteq mock. Did we optimize them?

File: panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+TEST_F(TestRoboteqDriver, ResetRoboteqScriptSDOTimeoutReset)
+{
+  BootRoboteqDriver();
+
+  roboteqs_mock_->GetDriver()->SetOnWriteWait<std::uint8_t>(0x2018, 0, 100000);
+  EXPECT_TRUE(panther_utils::test_utils::IsMessageThrown<std::runtime_error>(
+    [&]() { roboteq_driver_->ResetScript(); }, "SDO protocol timed out"));
+
+  roboteqs_mock_->GetDriver()->SetOnWriteWait<std::uint8_t>(0x2018, 0, 0);
+  ASSERT_NO_THROW(roboteq_driver_->ResetScript());
```
Comment (pawelirh) [inline panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp]: Default choice should be EXPECT, expecially at the end of test case.

File: panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+}
+
+// OnCanError/OnHeartbeatTimeout isn't tested, because it reacts to lower-level CAN errors (CRC),
+// which are hard to simulate, but it would be nice to add it
+
+int main(int argc, char ** argv)
+{
+  testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+}
```
Comment (pawelirh) [inline panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp]:
```suggestion
int main(int argc, char ** argv)
{
  testing::InitGoogleTest(&argc, argv);
  
  auto result = RUN_ALL_TESTS();

  return result;
}
```

File: panther_hardware_interfaces/test/utils/roboteqs_mock.hpp [hunk 8]
```diff
@@ -267,7 +267,7 @@ class RoboteqsMock
   ~RoboteqsMock() {}
 
   /**
-   * @brief Starts CAN communication and creates two simulated Roboteqs, that publish PDOs with set
+   * @brief Starts CAN communication and creates a simulated Roboteq, that publish PDOs with set
```
Comment (pawelirh) [inline panther_hardware_interfaces/test/utils/roboteqs_mock.hpp]: If we mock one roboteq we need to rename a file, class and some descriptions.

File: panther_hardware_interfaces/test/utils/roboteqs_mock.hpp
Comment (pawelirh) [inline panther_hardware_interfaces/test/utils/roboteqs_mock.hpp:1]: How about using first mock/fake/... etc. as a file and class name so we can later group them by types?

**Review by pawelirh — CHANGES_REQUESTED** [review]

File: panther_hardware_interfaces/src/panther_system/motors_controller/roboteq_driver.cpp (hunk 1 above)
Comment (KmakD) [reply panther_hardware_interfaces/src/panther_system/motors_controller/roboteq_driver.cpp:36]: I'll move it. Then name of the library is actually "CANopen" so I would stick with that.

File: panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp (hunk 3 above)
Comment (KmakD) [reply panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp]: feedback, want it changed to sth lie driver_state_motor_1?

File: panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp (hunk 2 above)
Comment (KmakD) [reply panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp]: Copied it from different tests, no idea why it was done this way. Fixed

File: panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp (hunk 4 above)
Comment (KmakD) [reply panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp]: The comment is misleading, the 10 referred to 10ms period, fixed.

File: panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp (hunk 5 above)
Comment (KmakD) [reply panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp:231]: About 5s. This is based on roboteq_mock publishing frequencies for updating driver state. 
The most time consuming are boots, which takes about 100-150 ms for each test.

File: panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp (hunk 3 above)
Comment (pawelirh) [reply panther_hardware_interfaces/test/unit/panther_system/motors_controller/test_roboteq_driver.cpp]: Yep, I'd rename since it's not intuitive 👍🏼

**Pull request merged**